### PR TITLE
Added detection rule for DllRegisterServer export function load

### PR DESCRIPTION
**Pull request by netgrain**
### Summary of the Pull Request

Adds a detection rule for DLL export function load via RunDLL32 (explicit) and RegSvr32 (implicit). 

Tested against known samples and in a real-world environment.

e.g. https://thedfirreport.com/2023/08/28/html-smuggling-leads-to-domain-wide-ransomware/

### Changelog

new: Regsvr32.EXE Calling of DllRegisterServer Export Function Implicitly
new: Rundll32.EXE Calling DllRegisterServer Export Function Explicitly 

### Example Log Event
 
 - https://www.virustotal.com/gui/file/288fc4f954f98d724e6fab32a89477943df5c0e9662cb199a19b90ae0c63aebe/detection 
 - https://www.virustotal.com/gui/file/d3db55cd5677b176eb837a536b53ed8c5eabbfd68f64b88dd083dc9ce9ffb64e/details

```
{
  "Timestamp": "2023-08-31T06:20:28.5548387Z",
  "DeviceId": "<DeviceId>",
  "DeviceName": "<DeviceName>",
  "ActionType": "ProcessCreated",
  "FileName": "rundll32.exe",
  "FolderPath": "C:\\Windows\\System32\\rundll32.exe",
  "SHA1": "dd399ae46303343f9f0da189aee11c67bd868222",
  "SHA256": "b53f3c0cd32d7f20849850768da6431e5f876b7bfa61db0aa0700b02873393fa",
  "MD5": "ef3179d498793bf4234f708d3be28633",
  "FileSize": 71680,
  "ProcessVersionInfoCompanyName": "Microsoft Corporation",
  "ProcessVersionInfoProductName": "Microsoft® Windows® Operating System",
  "ProcessVersionInfoProductVersion": "10.0.19041.746",
  "ProcessVersionInfoInternalFileName": "rundll",
  "ProcessVersionInfoOriginalFileName": "RUNDLL32.EXE",
  "ProcessVersionInfoFileDescription": "Windows host process (Rundll32)",
  "ProcessId": 6300,
  "ProcessCommandLine": "rundll32.exe  1.dll, DllRegisterServer",
  "ProcessIntegrityLevel": "High",
  "ProcessTokenElevation": "TokenElevationTypeFull",
  "ProcessCreationTime": "2023-08-31T14:20:23.5918460Z",
  "AccountDomain": "<AccountDomain>",
  "AccountName": "<AccountName>",
  "AccountSid": "<AccountSid>",
  "AccountUpn": "",
  "AccountObjectId": "",
  "LogonId": 0,
  "InitiatingProcessAccountDomain": "<InitiatingProcessAccountDomain>",
  "InitiatingProcessAccountName": "<InitiatingProcessAccountName>",
  "InitiatingProcessAccountSid": "<InitiatingProcessAccountSid>",
  "InitiatingProcessAccountUpn": "",
  "InitiatingProcessAccountObjectId": "",
  "InitiatingProcessLogonId": 0,
  "InitiatingProcessIntegrityLevel": "High",
  "InitiatingProcessTokenElevation": "TokenElevationTypeFull",
  "InitiatingProcessSHA1": "f1efb0fddc156e4c61c5f78a54700e4e7984d55d",
  "InitiatingProcessSHA256": "b99d61d874728edc0918ca0eb10eab93d381e7367e377406e65963366c874450",
  "InitiatingProcessMD5": "8a2122e8162dbef04694b9c3e0b6cdee",
  "InitiatingProcessFileName": "cmd.exe",
  "InitiatingProcessFileSize": 289792,
  "InitiatingProcessVersionInfoCompanyName": "Microsoft Corporation",
  "InitiatingProcessVersionInfoProductName": "Microsoft® Windows® Operating System",
  "InitiatingProcessVersionInfoProductVersion": "10.0.19041.746",
  "InitiatingProcessVersionInfoInternalFileName": "cmd",
  "InitiatingProcessVersionInfoOriginalFileName": "Cmd.Exe",
  "InitiatingProcessVersionInfoFileDescription": "Windows Command Processor",
  "InitiatingProcessId": 1248,
  "InitiatingProcessCommandLine": "\"cmd.exe\" ",
  "InitiatingProcessCreationTime": "2023-08-31T14:18:44.0702353Z",
  "InitiatingProcessFolderPath": "c:\\windows\\system32\\cmd.exe",
  "InitiatingProcessParentId": 7848,
  "InitiatingProcessParentFileName": "svchost.exe",
  "InitiatingProcessParentCreationTime": "2023-08-29T23:14:45.6713227Z",
  "InitiatingProcessSignerType": "OsVendor",
  "InitiatingProcessSignatureStatus": "Valid",
  "ReportId": 22813,
  "AppGuardContainerId": "",
  "AdditionalFields": ""
}
```

### SigmaHQ Rule Creation Conventions

- If your PR adds new rules, please consider following and applying these [conventions](https://github.com/SigmaHQ/sigma-specification/blob/main/sigmahq/sigmahq_conventions.md)

### CODE REVIEW

**Comment by nasbench** [comment]
This unfortunately won't work with rundll32 as the "DllRegisterServer" export is called from all paths around in legitimate form. That's why the rule in the repo looks only for renamed rundll32. Because the other way is too prone to FP

As for the regsvr32 selection. I don't understand what you're trying to do. Regsvr32 would call the export by default if nothing is provided so the "/s" isn't required and what do are you implying by using the empty double quotes?

**Comment by netgrain** [comment]
> This unfortunately won't work with rundll32 as the "DllRegisterServer" export is called from all paths around in legitimate form. That's why the rule in the repo looks only for renamed rundll32. Because the other way is too prone to FP. We can switch the rundll32 to a hunting rule.

I could be wrong, but pivoting in high telemetry datasets produced by various EDR-platforms and log-sources, it does not seem to be that common to call DllRegisterServer outside of Program Files* path. e.g. EventID 4688 or Sysmon id 1. But aware of some legitimate exceptions (as evident by e.g. [VT](https://www.virustotal.com/gui/search/type%253Apedll%2520exports%253ADllRegisterServer%2520positives%253A0%252B%2520have%253Acomments%2520filename%253A*.dll/files)), hence the Medium sev.

Perfectly fine with it re-cataloged as a hunting rule as opposed to detection rule.

> As for the regsvr32 selection. I don't understand what you're trying to do. Regsvr32 would call the export by default if nothing is provided so the "/s" isn't required and what do are you implying by using the empty double quotes?

In some malware cases the calling is as straightforward as 'regsvr32.exe /s "<dll>"', which again if I'm not mistaken, is somewhat unique for maliciousness when no other arguments are provided (and also reduces the need for further filtering e.g. if one would look for only the Image initiation). Sure there are better ways of expressing but, but what I had (and tested successfully) off the bat.

Altogether rundll32 and regsvr32 could be split into different detection/hunting logic rules, but makes more sense to me   as the execution intention is likely to be the same. A matter of preferences I guess :-)

**Comment by nasbench** [comment]
> I could be wrong, but pivoting in high telemetry datasets produced by various EDR-platforms and log-sources, it does not seem to be that common to call DllRegisterServer outside of Program Files* path. e.g. EventID 4688 or Sysmon id 1. But aware of some legitimate exceptions (as evident by e.g. [VT](https://www.virustotal.com/gui/search/type%253Apedll%2520exports%253ADllRegisterServer%2520positives%253A0%252B%2520have%253Acomments%2520filename%253A*.dll/files)), hence the Medium sev.
> 
> Perfectly fine with it re-cataloged as a hunting rule as opposed to detection rule.

I've tested the logic before and a lot of installers and random software generated that event. The maliciousness is undeniable but since there is no filtering or specifics, its best suited as a hunting :)

> In some malware cases the calling is as straightforward as 'regsvr32.exe /s ""', which again if I'm not mistaken, is somewhat unique for maliciousness when no other arguments are provided (and also reduces the need for further filtering e.g. if one would look for only the Image initiation). Sure there are better ways of expressing this, but what I had (and tested successfully) off the bat.

I agree with the maliciousness/susp aspect. That kind of exec is interesting to look at. If you could please provide a public reference or a log example for this, that would be great. As from a future reader pov the intention its unclear.


> Altogether rundll32 and regsvr32 could be split into different detection/hunting logic rules, but makes more sense to me as the execution intention is likely to be the same. A matter of preferences I guess :-)

We usually tend to split rules that have different appetite of FP and suspiciousness level. To make it easier for the analyst to tune and respond and so that the metadata of the rule is clear.

**Comment by netgrain** [comment]
> I've tested the logic before and a lot of installers and random software generated that event. The maliciousness is undeniable but since there is no filtering or specifics, its best suited as a hunting :)

As you want :+1: 

> I agree with the maliciousness/susp aspect. That kind of exec is interesting to look at. If you could please provide a public reference or a log example for this, that would be great. As from a future reader pov the intention its unclear.

Here you go https://www.virustotal.com/gui/file/288fc4f954f98d724e6fab32a89477943df5c0e9662cb199a19b90ae0c63aebe/detection

Typical behavior for a lot of non-customised CS stagers.

> We usually tend to split rules that have different appetite of FP and suspiciousness level. To make it easier for the analyst to tune and respond and so that the metadata of the rule is clear.

Sure. The FP ratio and suspiciousness level were the same on the dataset I used, but makes sense to split to accommodate future filtering and organisational needs.

I'll update soon-ish!

**Comment by netgrain** [comment]
Updated as requested, @nasbench

File: rules-threat-hunting/windows/process_creation/proc_creation_win_regsvr32_execution.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    - attack.t1218
+logsource:
+    category: process_creation
+    product: windows
+detection:
+    selection_image:
+        - Image|endswith: '\regsvr32.exe'
+        - OriginalFileName: 'REGSVR32.EXE'
+    selection_cmdline:
+        CommandLine|contains: ' /s ""'
```
Comment (nasbench) [inline rules-threat-hunting/windows/process_creation/proc_creation_win_regsvr32_execution.yml:25]: I still don't understand the logic behind this empty strings. I spend some time looking for an example but couldn't find one leveraging an empty string. If the empty string is there how is the DLL provided? 

Can you provide an example CLI of what you found in the wild?

The VT link you provided only links to a DLL with the `DllRegisterServer` export.

**Review by nasbench — COMMENTED** [review]
Hey @netgrain sorry for the long delay on this one.

Just need your input on one rule before we can merge this PR. If you may :)

File: rules-threat-hunting/windows/process_creation/proc_creation_win_regsvr32_execution.yml (hunk 1 above)
Comment (nasbench) [reply rules-threat-hunting/windows/process_creation/proc_creation_win_regsvr32_execution.yml:25]: cc @netgrain any info you can provide?

Comment (netgrain) [reply rules-threat-hunting/windows/process_creation/proc_creation_win_regsvr32_execution.yml]: Your right @nasbench. 

Minor double quotes typo, which now is revised per commit to capture all subsequent pathing.

**Comment by nasbench** [comment]
Just some quick notes for the changes introduced

- I added some additional filters to account for FP found in test data.
- For the regsvr32 rule added common flags to keep the logic of the rule. Which is `calls to regsvr32 with /s and no other flags from uncommon location`

Thanks for the contribution once again @netgrain

File: rules-threat-hunting/windows/process_creation/proc_creation_win_regsvr32_dllregisterserver_exec.yml [hunk 2]
```diff
@@ -0,0 +1,54 @@
+title: Indirect Export Function DllRegisterServer Called By Regsvr32.EXE
```
Comment (phantinuss) [inline rules-threat-hunting/windows/process_creation/proc_creation_win_regsvr32_dllregisterserver_exec.yml:1]:
```suggestion
title: Regsvr32.EXE Calling of DllRegisterServer Export Function Implicitly
```

File: rules-threat-hunting/windows/process_creation/proc_creation_win_regsvr32_dllregisterserver_exec.yml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        # Note: We filter other flags to keep the logic of the rule
+        CommandLine|contains:
+            - ' /i:'
+            - '/U '
+    filter_main_rpcproxy:
+        ParentCommandLine|endswith: ':\Windows\System32\RpcProxy\RpcProxy.dll'
+        CommandLine: 'regsvr32 /s rpcproxy.dll'
+    condition: all of selection_* and not 1 of filter_main_*
+falsepositives:
+    - Is legitimately used as part of application install, but less likely from e.g. temporary paths.
```
Comment (phantinuss) [inline rules-threat-hunting/windows/process_creation/proc_creation_win_regsvr32_dllregisterserver_exec.yml:53]:
```suggestion
    - Legitimate usage as part of application installation, but less likely from e.g. temporary paths.
```

File: rules-threat-hunting/windows/process_creation/proc_creation_win_rundll32_dllregisterserver.yml [hunk 4]
```diff
@@ -0,0 +1,38 @@
+title: Export Function DllRegisterServer Called By Rundll32.EXE
```
Comment (phantinuss) [inline rules-threat-hunting/windows/process_creation/proc_creation_win_rundll32_dllregisterserver.yml:1]:
```suggestion
title: Rundll32.EXE Calling DllRegisterServer Export Function Explicitly 
```

File: rules-threat-hunting/windows/process_creation/proc_creation_win_rundll32_dllregisterserver.yml [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+id: d81a9fc6-55db-4461-b962-0e78fea5b0ad
+related:
+    - id: 2569ed8c-1147-498a-9b8c-2ad3656b10ed # Renamed rundll32
+      type: similar
+status: experimental
+description: |
+    Detects when the DLL export function 'DllRegisterServer' is called in the commandline by Rundll32 (explicitly where the DLL is located in a non-standard path.
```
Comment (phantinuss) [inline rules-threat-hunting/windows/process_creation/proc_creation_win_rundll32_dllregisterserver.yml:8]:
```suggestion
    Detects when the DLL export function 'DllRegisterServer' is called in the commandline by Rundll32 explicitly where the DLL is located in a non-standard path.
```

File: rules-threat-hunting/windows/process_creation/proc_creation_win_rundll32_dllregisterserver.yml [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    filter_main_legit_paths:
+        CommandLine|contains:
+            - ':\Program Files (x86)'
+            - ':\Program Files\'
+            - ':\Windows\System32\'
+            - ':\Windows\SysWOW64\'
+    condition: all of selection_* and not 1 of filter_main_*
+falsepositives:
+    - Is legitimately used as part of application install, but less likely from e.g. temporary paths.
+    - Not every instance is considered malicious, but it will capture malicious use of it.
```
Comment (phantinuss) [inline rules-threat-hunting/windows/process_creation/proc_creation_win_rundll32_dllregisterserver.yml:37]:
```suggestion
    - Legitimate usage as part of application installation, but less likely from e.g. temporary paths.
    - Not every instance is considered malicious, but this rule will capture the malicious usages.
```

**Review by phantinuss — APPROVED** [review]

**Pull request merged**